packages/ipfs-daemon/src/create-ipfs.ts

```diff
@@ -1,87 +1,91 @@
-  const swarmPort = await getPort()
-  const apiPort = await getPort()
-  const gatewayPort = await getPort()
-  const defaultConfig: Partial<Options> = {
-    ipld: { codecs: [dagJose] },
-    config: {
-      Pubsub: {
-        Enabled: true,
-      },
-      Addresses: {
-        Swarm: [`/ip4/127.0.0.1/tcp/${swarmPort}`],
-        Gateway: `/ip4/127.0.0.1/tcp/${gatewayPort}`,
-        API: `/ip4/127.0.0.1/tcp/${apiPort}`,
-      },
-      Bootstrap: [],
-    },
-  }
+const ipfsHttpModule = {
+  create: (ipfsEndpoint: string) => {
+    return ipfsClient.create({
+      url: ipfsEndpoint,
+      ipld: { codecs: [dagJose] },
+    })
+  },
+}
 
-  return mergeOptions(defaultConfig, override) as Options
+const ipfsOptions = {
+  ipld: { codecs: [dagJose] },
+  config: {
+    Pubsub: {
+      Enabled: true,
+    },
+    Bootstrap: [],
+  },
 }
 
-/**
- * Create an IPFS instance
- * @param overrideConfig - IFPS config for override
- */
-export async function createIPFS(overrideConfig: Partial<Options> = {}): Promise<IpfsApi> {
-  const flavor = process.env.IPFS_FLAVOR
-  if (flavor && flavor.toLowerCase() == 'js') {
-    return createJsIPFS(overrideConfig)
-  } else {
-    return createGoIPFS(overrideConfig)
-  }
+const createFactory = () => {
+  return Ctl.createFactory(
+    {
+      ipfsHttpModule,
+      disposable: true,
+      ipfsOptions,
+    },
+    {
+      go: {
+        ipfsBin: path(),
+      },
+    }
+  )
 }
 
-/**
- * Create go-ipfs instance
- * @param overrideConfig - IFPS config for override
- */
-async function createGoIPFS(overrideConfig: Partial<Options> = {}): Promise<IpfsApi> {
-  const ipfsOptions = await goIpfsConfig(overrideConfig)
+async function createGoIpfsOptions(override: Partial<Options> = {}): Promise<Options> {
+  const tmpFolder = await tmp.dir({ unsafeCleanup: true })
+  const port = await getPort()
+  const swarmPort = await getPort()
+  const apiPort = await getPort()
+  const gatewayPort = await getPort()
 
-  const ipfsd = await Ctl.createController({
-    ipfsHttpModule: ipfsHttp,
-    ipfsBin: path(),
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    // @ts-ignore ipfsd-ctl uses own type, that is _very_ similar to Options from ipfs-core
-    ipfsOptions: ipfsOptions,
-    disposable: true,
-  })
-  return ipfsd.api
+  return mergeOptions(
+    {
+      repo: `${tmpFolder.path}/ipfs${port}/`,
+      config: {
+        Addresses: {
+          Swarm: [`/ip4/127.0.0.1/tcp/${swarmPort}`],
+          Gateway: `/ip4/127.0.0.1/tcp/${gatewayPort}`,
+          API: `/ip4/127.0.0.1/tcp/${apiPort}`,
+        },
+      },
+    },
+    override
+  )
 }
 
 /**
  * Create js-ipfs instance
  * @param overrideConfig - IFPS config for override
  */
-async function createJsIPFS(overrideConfig: Record<string, unknown> = {}): Promise<IpfsApi> {
+async function createJsIpfs(overrideConfig: Record<string, unknown> = {}): Promise<IpfsApi> {
```

Comment:
All of these helpers are used for testing only. HTTP server for in-proess js-ipfs is not required here, as Ceramic-IPFS communication happens via vanilla JS calls.

Reply:
oh, I didn't realize these were testing only.  Is this not what is used when you run the ipfs-daemon package directly?

Reply:
The next work-block makes IpfsDaemon spawn go-ipfs instance for ceramic cli. For now, this _all_ is just for testing.

Reply:
And, we are not going to use IpfsDaemon on its own, like in Docker image or anything.

Reply:
2.0 release is going to break that package indeed. That was always the case. See discussion starting at https://discord.com/channels/484729862368526356/669564901261443125/930583097735540737. TLDR: We are on vanilla go-ipfs, so no plugins are expected. That’s why bundled go-ipfs is only used for local development. On cloud we require an external go-ipfs node.

Reply:
right, but what about the users who are using ipfs-daemon to launch a **js**-ipfs process?  Why would we break them?  Why not continue to let the ipfs-daemon package launch js-ipfs, then then users can switch to vanilla go-ipfs manually when they want?

The 2.0 upgrade is already going to be complex and breaking without forcing devs to switch to go-ipfs at the exact same time before they can even get their app back online.

Reply:
"what about the users who are using ipfs-daemon to launch a js-ipfs process" - Based on my understanding, this kind of users is the absolute majority, since it is the easiest, most convenient, and recommended way of running js-ipfs with all the features we need. And, these people are the target of our intended breaking change. And, the whole point of this release to combine the most painful breaking changes together - that includes mandatory use of go-ipfs.

Reply:
Right, but it's still valuable to break changes down into smaller pieces that can be done independently whenever possible.  The larger the set of things that needs to change all at once the more risk there is and the longer an app is likely to be offline during the upgrade.

If there's some reason we *need* to break the ipfs-daemon package and force our users to upgrade their ceramic version, upgrade their node version, switch to esm modules, and switch ipfs deployments all at once, then so be it.  All I'm saying is that if we can break any of those pieces out and allow them to be done independently of the others, that's a benefit for our users.  Users who want to do it all at once still can, but those that would rather do each piece one at a time would have the option to do so.

Is there a reason we need to break the ipfs-daemon package?  Can we not just keep it working the way it is today when run directly?

Reply:
go-ipfs enables DHT and proper peer discovery. With this, there is no need to maintain peerlist. js-ipfs have no such luxury. If our goal is open mainnet, then dropping of js-ipfs is necessary.

Reply:
eventually yes we'll want to drop support for js-ipfs, but not on day 0.  I imagined us continuing to support both in parallel for at least a few months to give people time to move over.

You still haven't answered my question - is there a reason we *have* to break this package?  How much extra work would it be to keep the ipfs-daemon package working as it does today?  If it's on the order of a day or two, I think we should do it (although that need not hold up this PR, just the eventual release).  If it's over a week or two of work then we'd need to reconsider, but that would be surprising to me given that it works currently.

Reply:
FYI @oed - we're discussing whether the ipfs-daemon package should continue to work for launching a js-ipfs process as it does today when ceramic 2.0 is released, or if the ceramic 2.0 release should remove all support for running a js-ipfs node and force our users to switch to go-ipfs immediately.

Reply:
Hm, why would ipfs-daemon need to support js-ipfs? If users want to continue using js-ipfs they can use the old release of ipfs-daemon or just run js-ipfs directly (mod https://github.com/ipfs/js-ipfs/pull/3989).
I think it's more important that our testsuite support both js and go ipfs for the time being.

Reply:
Okay guess I'm overruled.  I hadn't really thought about the fact that users could upgrade their core ceramic packages separately from when they upgrade their ipfs-daemon package for the ipfs process they are running.  That makes me feel better about this.